game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java

```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
-                hitsLeft,
-                null,
-                DiceType.ENGINE,
-                "Deciding which planes should die due to AA fire");
-        int pos = 0;
-        for (final int element : hitRandom) {
-          pos += element;
-          final Unit unitHit = airSplit.getSecond().remove(pos % airSplit.getSecond().size());
-          if (allowMultipleHitsPerUnit
-              && (Collections.frequency(finalCasualtyDetails.getDamaged(), unitHit)
-                  < (getTotalHitpointsLeft(unitHit) - 1))) {
-            finalCasualtyDetails.addToDamaged(unitHit);
-          } else {
-            finalCasualtyDetails.addToKilled(unitHit);
-          }
-        }
+      // There is somehow more guaranteed hit groups than hits. This currently only happens
+      // with multi hp targets and damageable AA shots.
+
+      // pull out one unit from each guaranteed hit and then randomly pick the hits from those
+      final List<Unit> guaranteedHitUnits = new ArrayList<>();
+      for (final List<Unit> group : targetGroups.guaranteedHitGroups) {
+        guaranteedHitUnits.add(group.get(0));
       }
-    }
 
-    // double check
-    if (finalCasualtyDetails.size() != dice.getHits()) {
-      throw new IllegalStateException(
-          "wrong number of casualties, expected:" + dice + " but got: " + finalCasualtyDetails);
+      return buildCasualtyDetails(
+          availableTargets, findRandomTargets(guaranteedHitUnits, bridge, dice.getHits()));
     }
-    return finalCasualtyDetails;
   }
 
-  private static CasualtyDetails calculateAaCasualties(
-      final List<Unit> availableTargets,
-      final AaPowerStrengthAndRolls unitPowerAndRollsMap,
-      final DiceRoll dice,
-      final IDelegateBridge bridge) {
+  /**
+   * Categorize the units and then split them up into groups of guaranteeHitGroupSize
+   *
+   * <p>Any group less than guaranteeHitGroupSize is added to the remainderUnits
+   */
+  private static LowLuckTargetGroups createGuaranteedLowLuckHitGroups(
+      final Collection<Unit> targets,
+      final DiceRoll diceRoll,
+      final AaPowerStrengthAndRolls unitPowerAndRollsMap) {
 
-    final CasualtyDetails finalCasualtyDetails = new CasualtyDetails();
-    final int hits = dice.getHits();
-    final Set<Integer> hitTargets;
-    if (unitPowerAndRollsMap.calculateTotalRolls() == availableTargets.size()
-        && hits < availableTargets.size()) {
-      // there is a roll for every target but not enough hits to kill all of the targets
-      // so no need to get a random set of units since all units will either have a hit
-      // or miss roll
-      final List<Die> rolls = dice.getRolls();
-      hitTargets = new HashSet<>();
-      for (int i = 0; i < rolls.size(); i++) {
-        if (rolls.get(i).getType() == DieType.HIT) {
-          hitTargets.add(i);
-        }
+    final int guaranteeHitGroupSize =
+        calculateGuaranteeLowLuckHitGroupSize(targets, diceRoll, unitPowerAndRollsMap);
```

Comment:
Is this a place where we could early terminate and avoid processing on empty lists?

EG:
```
    final int guaranteeHitGroupSize =
        calculateGuaranteeLowLuckHitGroupSize(targets, diceRoll, unitPowerAndRollsMap);
   if(guaranteedHitGroupSize == 0) {
           return LowLuckTargetGroups.NO_GUARANTEED_HITS;
    }
```

Note, this comment was made out of order for other comments below, this could moot some of those suggestions.